shell/common/shell.cc

```diff
@@ -207,150 +273,84 @@ std::unique_ptr<Shell> Shell::CreateShellOnPlatformThread(
-    }
-
-    tonic::SetLogHandler(
-        [](const char* message) { FML_LOG(ERROR) << message; });
-
-    if (settings.trace_skia) {
-      InitSkiaEventTracer(settings.trace_skia);
-    }
-
-    if (!settings.trace_allowlist.empty()) {
-      std::vector<std::string> prefixes;
-      Tokenize(settings.trace_allowlist, &prefixes, ',');
-      fml::tracing::TraceSetAllowlist(prefixes);
-    }
-
-    if (!settings.skia_deterministic_rendering_on_cpu) {
-      SkGraphics::Init();
-    } else {
-      FML_DLOG(INFO) << "Skia deterministic rendering is enabled.";
-    }
-
-    if (settings.icu_initialization_required) {
-      if (settings.icu_data_path.size() != 0) {
-        fml::icu::InitializeICU(settings.icu_data_path);
-      } else if (settings.icu_mapper) {
-        fml::icu::InitializeICUFromMapping(settings.icu_mapper());
-      } else {
-        FML_DLOG(WARNING) << "Skipping ICU initialization in the shell.";
-      }
-    }
-  });
-}
-
 std::unique_ptr<Shell> Shell::Create(
+    const PlatformData platform_data,
     TaskRunners task_runners,
     Settings settings,
     const Shell::CreateCallback<PlatformView>& on_create_platform_view,
     const Shell::CreateCallback<Rasterizer>& on_create_rasterizer) {
-  return Shell::Create(std::move(task_runners),                    //
-                       PlatformData{/* default platform data */},  //
-                       std::move(settings),                        //
-                       std::move(on_create_platform_view),         //
-                       std::move(on_create_rasterizer)             //
-  );
-}
-
-std::unique_ptr<Shell> Shell::Create(
-    TaskRunners task_runners,
-    const PlatformData platform_data,
-    Settings settings,
-    Shell::CreateCallback<PlatformView> on_create_platform_view,
-    Shell::CreateCallback<Rasterizer> on_create_rasterizer) {
+  // This must come first as it initializes tracing.
   PerformInitializationTasks(settings);
-  PersistentCache::SetCacheSkSL(settings.cache_sksl);
 
   TRACE_EVENT0("flutter", "Shell::Create");
 
   auto vm = DartVMRef::Create(settings);
   FML_CHECK(vm) << "Must be able to initialize the VM.";
 
-  auto vm_data = vm->GetVMData();
-
-  return Shell::Create(std::move(task_runners),        //
-                       std::move(platform_data),       //
-                       std::move(settings),            //
-                       vm_data->GetIsolateSnapshot(),  // isolate snapshot
-                       on_create_platform_view,        //
-                       on_create_rasterizer,           //
-                       std::move(vm),                  //
-                       CreateEngine);
+  // Use the Isolate snapshot provided by the settings (in case the VM was
+  // already running); otherwise use the initial Isolate snapshot that the VM
+  // was created with.
+  fml::RefPtr<const DartSnapshot> isolate_snapshot =
+      DartSnapshot::IsolateSnapshotFromSettings(settings);
```

Comment:
This is the fix, right?  I think ideally in the case of the Spawn, `IsolateSnapshotFromSettings` would return null right away.  I looked over the code briefly and it looks kind of involved and I'm a bit concerned that it won't return null in practice.  That will make us incur unnecessary cpu cycles during spawning where grabbing the isolate snapshot from the vm would be faster, right?  If that is the case, is there another conditional we could put here to decide between reading from settings versus the vm, one that is quicker?

Reply:
This codepath isn't taken in the Spawn case, is it?  This is inside of Shell::Create, not Shell::CreateWithSnapshot.  So `IsolateSnapshotFromSettings` won't be called in the Spawn case.

Shell::Create is used when launching a new Shell from scratch, so it should always use the one from the settings since the settings might contain a new isolate snapshot to be used.  The one exception is the very first Shell in the process -- in that case the settings are indeed parsed twice in a wasteful manner (once inside of DartVMRef::Create and then once again right here).

I can add a check if the VM is running and skip this call if it's the very first Shell, but that "is VM running check" is not synchronized at all - a 2nd Shell start on another thread could cause a false-positive and then we would load the wrong snapshot!

Reply:
Okay, sounds good to me!